### PR TITLE
task 6 GET/api/articles/:article_id/comments

### PR DESCRIPTION
**Pull request by holyseo**

### CODE REVIEW

File: __tests__/app.test.js [hunk 1]
```diff
@@ -91,7 +91,30 @@ describe("GET/api/articles", () => {
...
+        comments.forEach((comment) => {
+          expect(comment.article_id).toBe(1);
+          expect(comment).toMatchObject({
+            comment_id: expect.any(Number),
+            votes: expect.any(Number),
+            created_at: expect.any(String),
+            author: expect.any(String),
+            body: expect.any(String),
+            article_id: expect.any(Number),
+          });
```
Comment (ghost) [inline __tests__/app.test.js:139]: You are missing several tests here. In this test you aren't checking that the order returned is correct.

You are also missing:

- an article with no comments, what status code would this be?
- an article that doesn't exist, eg 99
- an article id that is invalid

File: endpoints.json [hunk 2]
```diff
@@ -28,5 +28,8 @@
   },
   "GET /api/articles/:article_id": {
     "description": "serves an array of an article found by article_id"
+  },
+  "GET /api/articles/:article_id/comments": {
+    "description": "serves an array of comments for an article with a specific article_id"
```
Comment (ghost) [inline endpoints.json]: Where is the rest of this? This file should contain much more information per end point

File: models/comments.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const db = require("../db/connection");
+
+exports.selectCommentsByArticleId = (id) => {
+  return db
+    .query(
+      `SELECT * FROM comments where article_id = $1 ORDER BY created_at DESC`,
+      [id]
+    )
+    .then(({ rows }) => {
+      console.log(rows);
```
Comment (ghost) [inline models/comments.js]: Make sure to remove console logs!

**Review by ghost — CHANGES_REQUESTED** [review]
This is almost there but missing quite a few tests. Take a look at my comment and please address then resubmit.

Looks like you also aren't testing that your endpoints.json contains the updated data, not that I can see in this commit anyway.

File: __tests__/app.test.js (hunk 1 above)
Comment (holyseo) [reply __tests__/app.test.js:139]: I have added a few more tests based on your suggestion.

File: endpoints.json (hunk 2 above)
Comment (holyseo) [reply endpoints.json]: More information added in endpoint.json file accordingly.

File: models/comments.js (hunk 3 above)
Comment (holyseo) [reply models/comments.js]: Thanks!

File: __tests__/app.test.js [hunk 4]
```diff
@@ -90,8 +106,46 @@ describe("GET/api/articles", () => {
...
+        });
+        expect(comments).toBeSortedBy("created_at", { descending: true });
+      });
+  });
+  it("status 404: responds with an error message when an article does not exist", () => {
+    return request(app)
+      .get("/api/articles/99/comments")
+      .expect(404)
+      .then(({ body }) => {
+        expect(body.msg).toBe("no comments found");
```
Comment (ghost) [inline __tests__/app.test.js]: You still aren't testing for an invalid article or an article with no comments. There is also no sort order on your comments. The full endpoint /api/:article_id/comments needs to be tested here, not just /api/:article_id 

You have the correct tests for your get all articles, however that isn't the endpoint you're working on. We need to see what happens when we request the comments from these articles.

File: error.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    res.status(err.status).send({ msg: err.msg });
+  }
+};
+
+exports.handleServerErrors = (err, req, res, next) => {
+  res.status(500).send({ msg: "internal server error" });
+};
+
+exports.handleErrors = (err, req, res, next) => {
+  if (err.status === 400) {
```
Comment (ghost) [inline error.js]: Does this actually fire?

File: error.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  }
+};
+
+exports.handleCustomErros = (err, req, res, next) => {
+  if (err.status) {
+    res.status(err.status).send({ msg: err.msg });
+  }
+};
+
+exports.handleServerErrors = (err, req, res, next) => {
```
Comment (ghost) [inline error.js]: Server errors should be at the bottom of your file. You aren't using next here so if this is not the right handler the functions below will never receive the error. Double check the error handling notes for this specifically.

File: models/comments.js [hunk 7]
```diff
@@ -0,0 +1,18 @@
+const db = require("../db/connection");
+
+exports.selectCommentsByArticleId = (id) => {
+  if (!id || isNaN(id)) {
```
Comment (ghost) [inline models/comments.js]: PSQL should be handling this. You have an error handler specifically for it, this should not be a custom reject.

File: models/comments.js [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+    return Promise.reject({ status: 400, msg: "article_id not found" });
+  }
+  return db
+    .query(
+      `SELECT * FROM comments where article_id = $1 ORDER BY created_at DESC ;`,
+      [id]
+    )
+    .then(({ rows }) => {
+      if (rows.length === 0) {
+        return Promise.reject({ status: 404, msg: "no comments found" });
```
Comment (ghost) [inline models/comments.js]: Is it an error if an article has no comments? What about a brand new article, no one would have commented on it yet. How is that an error?

**Review by ghost — CHANGES_REQUESTED** [review]
This isn't quite finished Steven. Please review the points and the missing tests then resubmit. Any questions please DM or use NCHelp as I don't have notifications on for you to reply to my comments.

File: __tests__/app.test.js [hunk 9]
```diff
@@ -137,15 +138,32 @@ describe("GET/api/articles/:article_id/comments", () => {
             article_id: expect.any(Number),
           });
         });
-        expect(comments).toBeSortedBy("created_at", { descending: true });
+      });
+  });
+  it("status 404: responds with an error message when an article_id is invalid", () => {
```
Comment (nc-poppy-m) [inline __tests__/app.test.js:143]: Is 404 the right error code for an invalid/bad request? Find out here (check the 400s): https://http.cat/

**Review by nc-poppy-m — APPROVED** [review]
Good work! Consider what error codes you need to use. Once this is updated, feel free to merge

**Pull request merged**